In [1]:
#  1.0 Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# 2.0 Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv('Resources/lending_data.csv')

# 2.1 Review the DataFrame
print(df.head())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  


In [3]:
# Get the dimensions of the DataFrame
dimensions = df.shape

# Print the dimensions
print("Number of rows:", dimensions[0])
print("Number of columns:", dimensions[1])

Number of rows: 77536
Number of columns: 8


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# 3.0 Separate the data into labels and features
# Separate the labels (y) and features (X)

  # Features from the remaining columns

# 3.1 Separate the y variable, the labels
y = df['loan_status']  # Labels from the "loan_status" column

# 3.2 Separate the X variable, the features
X = df.drop('loan_status', axis=1)


print("Shape of labels (y):", y.shape)
print("Shape of features (X):", X.shape)



Shape of labels (y): (77536,)
Shape of features (X): (77536, 7)


In [5]:
# 4.0 Review the y variable Series
# Review the y variable Series
print(y)

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64


In [6]:
# 5.0Review the X variable DataFrame
print(X)

       loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                 

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# 6.0Import the train_test_learn module
from sklearn.model_selection import train_test_split

# 6.1Split the data using train_test_split
# 6.2Assign a random_state of 1 to the function
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

# Display the shapes of the training and testing sets
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (58152, 7)
X_test shape: (19384, 7)
y_train shape: (58152,)
y_test shape: (19384,)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# 7.0Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# 7.1Instantiate the Logistic Regression model
# 7.2Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# 7.3Fit the model using training data

logistic_regression_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# 8.0 Make a prediction using the testing data
# Make predictions using the testing data
testing_predictions = logistic_regression_model.predict(X_test)

# Display the predictions
print(testing_predictions)


[0 0 0 ... 0 0 0]


In [10]:
# Display the first 5 predictions from the NumPy array
print(testing_predictions[:5])

[0 0 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# 9.0 Generate a confusion matrix for the model
conf_matrix = confusion_matrix(y_test, testing_predictions)
print("Confusion Matrix:")


print(conf_matrix)


Confusion Matrix:
[[18663   102]
 [   56   563]]


In [12]:
# 10.0 Print the classification report for the model
class_report = classification_report(y_test, testing_predictions)
print("\nClassification Report:")


print(class_report)



Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:
Based on the provided classification report and confusion matrix:

For the healthy loan label (0:

Precision:    .00
Recall:    0.99
F1-score     1.00
Support     18765
For the high-risk loan label (1):

Preci    ion: 0.85
R    call: 0.91
F1    score: 0.88

Support: 619
Overall, the logistic regression model performs very well in predicting both healthy loans and high-risk loans. It has high precision and recall values for both classes, indicating that it can effectively distinguish between the two classes. The accuracy of the model is 99%, which is excellent. The model's performance is further supported by the high F1-scores for both classes.

In summary, the logistic regression model demonstrates strong predictive performance for both healthy and high-risk loan labels based on the provided metrics.

---